File: java/src/aoc24/Day05.java
package aoc24;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

public class Day05 {

    public static void main(String[] args) throws IOException {
        Path inputPath = Path.of("resources", "aoc24", "day05.txt");
        List<String> input = Files.readAllLines(inputPath);

        String rulesStr = String.join("\n", input.subList(0, input.indexOf("")));
        String updatesStr = String.join("\n", input.subList(input.indexOf("") + 1, input.size()));

        long part1Result = solve(rulesStr, updatesStr);
        System.out.println("Part 1: " + part1Result);

        long part2Result = solvePart2(rulesStr, updatesStr);
        System.out.println("Part 2: " + part2Result);
    }

    public static List<List<Long>> parseRules(String rulesStr) {
        List<List<Long>> rules = new ArrayList<>();
        String[] lines = rulesStr.split("\n");
        for (String line : lines) {
            String[] parts = line.split("\\|");
            List<Long> rule = new ArrayList<>();
            rule.add(Long.parseLong(parts[0]));
            rule.add(Long.parseLong(parts[1]));
            rules.add(rule);
        }
        return rules;
    }

    public static List<Long> parseUpdate(String updateStr) {
        List<Long> update = new ArrayList<>();
        String[] parts = updateStr.split(",");
        for (String part : parts) {
            update.add(Long.parseLong(part));
        }
        return update;
    }

    public static List<List<Long>> parseUpdates(String updatesStr) {
        List<List<Long>> updates = new ArrayList<>();
        String[] lines = updatesStr.split("\n");
        for (String line : lines) {
            updates.add(parseUpdate(line));
        }
        return updates;
    }

    public static List<List<Long>> applicableRules(List<Long> update, List<List<Long>> rules) {
        List<List<Long>> appRules = new ArrayList<>();
        for (List<Long> rule : rules) {
            if (update.contains(rule.get(0)) && update.contains(rule.get(1))) {
                appRules.add(rule);
            }
        }
        return appRules;
    }

    public static boolean validateUpdate(List<Long> update, List<List<Long>> rules) {
        List<List<Long>> appRules = applicableRules(update, rules);
        for (List<Long> rule : appRules) {
            if (update.indexOf(rule.get(0)) >= update.indexOf(rule.get(1))) {
                return false;
            }
        }
        return true;
    }

    public static List<Long> middlePages(List<Long> update) {
        int len = update.size();
        int mid = len / 2;
        List<Long> result = new ArrayList<>();
        if (len % 2 == 1) {
            result.add(update.get(mid));
        } else {
            result.add(update.get(mid - 1));
            result.add(update.get(mid));
        }
        return result;
    }

    public static long sumMiddlePages(List<List<Long>> updates, List<List<Long>> rules) {
        long sum = 0;
        for (List<Long> update : updates) {
            if (validateUpdate(update, rules)) {
                List<Long> middle = middlePages(update);
                for (Long page : middle) {
                    sum += page;
                }
            }
        }
        return sum;
    }

    public static List<Long> correctUpdate(List<Long> update, List<List<Long>> rules) {
        List<Long> sortedUpdate = new ArrayList<>();
        List<Long> remainingUpdate = new ArrayList<>(update);

        while (!remainingUpdate.isEmpty()) {
            Long nextVal = remainingUpdate.remove(0);
            List<List<Long>> relevantRules = new ArrayList<>();
            for (List<Long> rule : rules) {
                if (rule.contains(nextVal)) {
                    relevantRules.add(rule);
                }
            }

            int correctIndex = 0;
            for (int i = 0; i <= sortedUpdate.size(); i++) {
                boolean validPlacement = true;
                for (List<Long> rule : relevantRules) {
                    List<Long> testUpdate = new ArrayList<>(sortedUpdate.subList(0, i));
                    testUpdate.add(nextVal);
                    testUpdate.addAll(sortedUpdate.subList(i, sortedUpdate.size()));

                    if (testUpdate.contains(rule.get(0)) && testUpdate.contains(rule.get(1))) {
                        if (testUpdate.indexOf(rule.get(0)) >= testUpdate.indexOf(rule.get(1))) {
                            validPlacement = false;
                            break;
                        }
                    }
                }
                if (validPlacement) {
                    correctIndex = i;
                    break;
                }
            }
            sortedUpdate.add(correctIndex, nextVal);
        }
        return sortedUpdate;
    }

    public static long sumMiddlePagesCorrected(List<List<Long>> updates, List<List<Long>> rules) {
        long sum = 0;
        for (List<Long> update : updates) {
            if (!validateUpdate(update, rules)) {
                List<Long> correctedUpdate = correctUpdate(update, rules);
                List<Long> middle = middlePages(correctedUpdate);
                for (Long page : middle) {
                    sum += page;
                }
            }
        }
        return sum;
    }

    public static long solve(String rulesStr, String updatesStr) {
        List<List<Long>> rules = parseRules(rulesStr);
        List<List<Long>> updates = parseUpdates(updatesStr);
        return sumMiddlePages(updates, rules);
    }

    public static long solvePart2(String rulesStr, String updatesStr) {
        List<List<Long>> rules = parseRules(rulesStr);
        List<List<Long>> updates = parseUpdates(updatesStr);
        return sumMiddlePagesCorrected(updates, rules);
    }
}

File: clojure/src/aoc24/day05.clj
(ns aoc24.day05
  (:require [clojure.java.io :as io]
            [clojure.string :as str]))

(def input (->> (slurp (io/resource "aoc24/day05.txt"))
                (str/split-lines)))

(defn parse-rule [rule-str]
  (->> (str/split rule-str #"\|")
       (map parse-long)))

(defn parse-rules [rules-str]
  (->> (str/split-lines rules-str)
       (map parse-rule)
       (set)))

(defn applicable-rules [update rules]
  (filter (fn [[x y]] (and (some #{x} update) (some #{y} update))) rules))

(defn validate-update [update rules]
  (let [app-rules (applicable-rules update rules)]
    (every? (fn [[x y]] (< (.indexOf update x) (.indexOf update y))) app-rules)))

(defn middle-pages [update]
  (let [len (count update)
        mid (quot len 2)]
    (if (odd? len)
      [(nth update mid)]
      [(nth update (dec mid)) (nth update mid)])))

(defn sum-middle-pages [updates rules]
  (->> updates
       (filter #(validate-update % rules))
       (mapcat middle-pages)
       (reduce + 0)))

(defn solve [rules-str updates-str]
  (let [rules (parse-rules rules-str)
        updates (->> (str/split-lines updates-str)
                     (map #(->> (str/split % #",") (map parse-long))))]
    (sum-middle-pages updates rules)))

(defn create-test-sequence [sorted-update idx new-val]
  (-> (subvec sorted-update 0 idx)
      (conj new-val)
      (into (subvec sorted-update idx))))

(defn valid-placement? [sorted-update relevant-rules idx next-val]
  (every? (fn [[x y]]
            (let [test-seq (create-test-sequence sorted-update idx next-val)]
              (if (and (some #{x} test-seq) (some #{y} test-seq))
                (< (.indexOf test-seq x) (.indexOf test-seq y))
                true)))
          relevant-rules))

(defn correct-update [update rules]
  (loop [sorted-update []
         remaining-update update]
    (if (empty? remaining-update)
      sorted-update
      (let [next-val (first remaining-update)
            relevant-rules (filter (fn [[x y]] (or (= x next-val) (= y next-val))) rules)
            positions (range (inc (count sorted-update)))]
        (when-let [insert-idx (first (filter #(valid-placement? sorted-update relevant-rules % next-val) positions))]
          (recur (create-test-sequence sorted-update insert-idx next-val)
                 (rest remaining-update)))))))

(defn sum-middle-pages-corrected [updates rules]
  (->> updates
       (remove #(validate-update % rules))
       (map #(correct-update % rules))
       (mapcat middle-pages)
       (reduce + 0)))

(defn solve-part-2 [rules-str updates-str]
  (let [rules (parse-rules rules-str)
        updates (->> (str/split-lines updates-str)
                     (map #(->> (str/split % #",") (map parse-long))))]
    (sum-middle-pages-corrected updates rules)))

(def rules-str "47|53\n97|13\n97|61\n97|47\n75|29\n61|13\n75|53\n29|13\n97|29\n53|29\n61|53\n97|53\n61|29\n47|13\n75|47\n97|75\n47|61\n75|61\n47|29\n75|13\n53|13")
(def updates-str "75,47,61,53,29\n97,61,53,29,13\n75,29,13\n75,97,47,61,53\n61,13,29\n97,13,75,29,47")

(solve rules-str updates-str)
(solve-part-2 rules-str updates-str)

(defn part-1
  "Run with bb -x aoc24.day05/part-1"
  [_]
  (let [[rules-lines _ updates-lines] (partition-by str/blank? input)
        rules-str (str/join "\n" rules-lines)
        updates-str (str/join "\n" updates-lines)]
    (prn (solve rules-str updates-str))))

(defn part-2
  "Run with bb -x aoc24.day05/part-2"
  [_]
  (let [[rules-lines _ updates-lines] (partition-by str/blank? input)
        rules-str (str/join "\n" rules-lines)
        updates-str (str/join "\n" updates-lines)]
    (prn (solve-part-2 rules-str updates-str))))

(comment

  (part-1 nil)
  (part-2 nil)

  'comment)
